### PR TITLE
Add unit test cases for Min and Max price for shipping methods

### LINKED ISSUE #8762: Add unit test for local pickup price selector
## Describe the bug

In #8623, we fixed the local pickup price in the shipping type selector and fixed the pickup price in the pickup options.

We need to add the unit test cases for the changes added in #8623. 

## Additional context

Slack: p1678881011738479-slack-C8X6Q7XQ U

### PR DESCRIPTION
**Pull request by tarunvijwani**
Fixes woocommerce/woocommerce-blocks#8762

In woocommerce/woocommerce-blocks#8623, we fixed the local pickup price in the shipping type selector and fixed the pickup price in the pickup options.

In this PR, we are adding the test cases for min and max values for shipping and local pickup options. 

### Changes in the PR:

- Added the unit test cases for `getShippingPrices` and `getLocalPickupPrices` helper functions. 

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Run `npm run test -- assets/js/base/utils/test/shipping-rates.ts` in terminal
2. Confirm all tests are passing. 

* [x] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### CODE REVIEW

File: assets/js/base/utils/test/shipping-rates.ts [hunk 1]
```diff
@@ -27,81 +35,224 @@ jest.mock( '@woocommerce/block-settings', () => ( {
...
+			meta_data: [],
+			selected: true,
+			currency_code: 'USD',
+			currency_symbol: '$',
+			currency_minor_unit: 2,
+			currency_decimal_separator: '.',
+			currency_thousand_separator: ',',
+			currency_prefix: '$',
+			currency_suffix: '',
+		},
```
Comment (nielslange) [inline assets/js/base/utils/test/shipping-rates.ts]: I wonder if we should introduce a helper function to avoid the duplications. Such a function could look like this:

```ts
const generateRate = (
  rate_id: string, 
  name: string, 
  price: string, 
  method_id: string, 
  selected: boolean, 
  instance_id: number
): typeof testPackage.shipping_rates[0] => {
  return {
    rate_id,
    name,
    description: '',
    delivery_time: '',
    price,
    taxes: '0',
    instance_id,
    method_id,
    meta_data: [],
    selected,
    currency_code: 'USD',
    currency_symbol: '$',
    currency_minor_unit: 2,
    currency_decimal_separator: '.',
    currency_thousand_separator: ',',
    currency_prefix: '$',
    currency_suffix: '',
  };
}
```

Rather than
```tsx
shipping_rates: [
		{
			rate_id: 'flat_rate:1',
			name: 'Flat rate',
			description: '',
			delivery_time: '',
			price: '10',
			taxes: '0',
			instance_id: 1,
			method_id: 'flat_rate',
			meta_data: [],
			selected: true,
			currency_code: 'USD',
			currency_symbol: '$',
			currency_minor_unit: 2,
			currency_decimal_separator: '.',
			currency_thousand_separator: ',',
			currency_prefix: '$',
			currency_suffix: '',
		},
		{ ... },
		{ ... },
		{ ... },
	],				
```
we could call
```tsx
shipping_rates: [
        generateRate('flat_rate:1', 'Flat rate', '10', 'flat_rate', true, 1),
		{ ... },
		{ ... },
		{ ... },
	],				
```

The helper function would automatically add all values that are used in all objects, and we would only have to cal the function with the values that are different in the various test objects.

While this is out of the scope of this PR, I recall that during our pairing session we talked about potentially creating a general file with these test objects, to share them with other files. I found the following sections, that could also benefit of such a helper function:

https://github.com/woocommerce/woocommerce-blocks/blob/a4fb03db8c53fe1101bfc5abc856ef0ce2716537/assets/js/base/components/cart-checkout/totals/shipping/test/index.tsx#L88

https://github.com/woocommerce/woocommerce-blocks/blob/a4fb03db8c53fe1101bfc5abc856ef0ce2716537/assets/js/blocks/checkout/inner-blocks/checkout-order-summary-block/test/block.js#L93

https://github.com/woocommerce/woocommerce-blocks/blob/a4fb03db8c53fe1101bfc5abc856ef0ce2716537/assets/js/blocks/checkout/inner-blocks/checkout-order-summary-block/test/block.js#L463

That said, touching the other files would exceed the scope of this PR. Nevertheless, I believe that we could benefit from introducing the suggested helper function, to reduce the duplicated code in this particular PR.

File: assets/js/base/utils/test/shipping-rates.ts [hunk 2]
```diff
@@ -27,81 +35,224 @@ jest.mock( '@woocommerce/block-settings', () => ( {
...
+	],
+};
+describe( 'Test Min and Max rates', () => {
+	it( 'correctly returns Min and Max rates for local pickup', () => {
```
Comment (nielslange) [inline assets/js/base/utils/test/shipping-rates.ts]: nitpick: 

```suggestion
	it( 'returns the lowest and highest rates when local pickup method is used', () => {
```

File: assets/js/base/utils/test/shipping-rates.ts [hunk 3]
```diff
@@ -27,81 +35,224 @@ jest.mock( '@woocommerce/block-settings', () => ( {
...
+				currency_suffix: '',
+			},
+		} );
 	} );
-	it( 'returns false for all rates if local pickup is disabled', () => {
-		// Attempt to assign to const or readonly variable error on next line is OK because it is mocked by jest
-		blockSettings.LOCAL_PICKUP_ENABLED = false;
-		const ratesToTest = [ 'flat_rate', 'local_pickup' ];
-		expect( hasCollectableRate( ratesToTest ) ).toBe( false );
+	it( 'correctly returns Min and Max rates for shipping', () => {
```
Comment (nielslange) [inline assets/js/base/utils/test/shipping-rates.ts]: nitpick:

```suggestion
	it( 'returns the lowest and highest rates when flat rate shipping method is used', () => {
```

**Review by nielslange — APPROVED** [review]
Thanks for working on this, @tarunvijwani. I'm pre-approving this PR, as it works as expected. 

However, I've identified a few things that could potentially be adjusted. Apart from the code changes, I suggested below, I also noticed that the unit tests only cover the happy path. I wonder if we should introduce units tests for edge cases? 🤔

**Comment by tarunvijwani** [comment]
Thank you so much, @nielslange, for the review! I have implemented all the changes you suggested and would be grateful if you could review it one more time.

**Comment by tarunvijwani** [comment]
Here is the ticket to refactor the test cases: https://github.com/woocommerce/woocommerce/issues/42366

**Review by nielslange — APPROVED** [review]
Thanks for applying my previous suggestions, @tarunvijwani. While looking at the updated file, I noticed two more minor things.

### 1. (Slightly) Redundant attribute values 

The helper function `generateRate()` has the following attributes:
```tsx
rateId: string,
name: string,
price: string,
instanceID: number,
methodId: string,
selected = false
```

When calling the helper function, the values of `name` and `methodId` look redundant. What do you think if we derive the value for `methodId` from `name` by lowercasing it and concatenating the strings with underscores? Maybe like this:
```tsx
const methodId = name.toLowerCase().split(' ').join('_');
```

Instead of calling 
```tsx
generateRate( 'flat_rate:1', 'Flat rate', '10', 1, 'flat_rate', false ),
```
we could then call
```tsx
generateRate( 'flat_rate:1', 'Flat rate', '10', 1, false ),
```

### 2. Default attribute values

I also noticed that when calling the helper function, we either don’t provide the attribute `selected`, or if we do, we’re using the value `false`. As this is the default value, we could remove is from all functions.

Applying both the previous and this suggestion, would change
```tsx
shipping_rates: [
	generateRate( 'flat_rate:1', 'Flat rate', '10', 1, 'flat_rate', false ),
	generateRate(
		'local_pickup:1',
		'Local pickup',
		'0',
		2,
		'local_pickup',
		false
	),
	generateRate(
		'local_pickup:2',
		'Local pickup',
		'10',
		3,
		'local_pickup',
		false
	),
	generateRate(
		'local_pickup:3',
		'Local pickup',
		'50',
		4,
		'local_pickup',
		false
	),
	generateRate( 'flat_rate:2', 'Flat rate', '50', 5, 'flat_rate', false ),
],
```
to
```tsx
shipping_rates: [
	generateRate( 'flat_rate:1', 'Flat rate', '10', 1 ),
	generateRate( 'local_pickup:1', 'Local pickup', '0', 2 ),
	generateRate( 'local_pickup:2', 'Local pickup', '10', 3 ),
	generateRate( 'local_pickup:3', 'Local pickup', '50', 4 ),
	generateRate( 'flat_rate:2', 'Flat rate', '50', 5 ),
],
```

What do you think about these two changes?

**Comment by tarunvijwani** [comment]
@nielslange Thank you so much! I have implemented your suggestions. The code looks amazing now. 🎉

**Comment by nielslange** [comment]
> @nielslange Thank you so much! I have implemented your suggestions. The code looks amazing now. 🎉

Yeah! Glad to hear that, @tarunvijwani. The code really looks great now. Much more compact and way better readable. 😀
